pkg/util/workqueue/timed_queue.go

```diff
@@ -39,24 +40,39 @@ func NewTimedWorkQueue(clock clock.Clock) *TimedWorkQueue {
 // Add adds the obj along with the current timestamp to the queue.
 func (q TimedWorkQueue) Add(obj interface{}) {
 	start := q.clock.Now()
-	item := timedWorkQueueItem{start, obj}
+	item := &timedWorkQueueItem{start, obj}
 	q.Type.Add(item)
 }
 
 // AddWithTimestamp is useful if the caller does not want to refresh the start
-// time when requeuing an item.
-func (q TimedWorkQueue) AddWithTimestamp(obj interface{}, timestamp time.Time) {
-	item := timedWorkQueueItem{timestamp, obj}
-	q.Type.Add(item)
+// time when requeuing an item. origin is the "origin" returned by Get().
+func (q TimedWorkQueue) AddWithTimestamp(origin, obj interface{}, timestamp time.Time) error {
```

Comment:
Make timedWorkQueueItem public.

Delete this method. Have Add check for an object being a *TimedWorkQueueItem before it wraps it in a TimedWorkQueueItem.